### PR TITLE
Use phar for infection and phpstan

### PR DESCRIPTION
**Pull request by BackEndTea**
This PR:

* Makes use of phars for infection and phpstan
* Does so by use of a `Makefile`
* Adds an `editorconfig`to make sure the Makefile uses tabs

Another option would be to make use of composer scripts, but i prefer a Makefile.

A downside of this is that `composer update` won't have an effect on these dependencies. Meaning one would have to manually update their version in the makefile.

Removed/changed dependencies:
For infection:
```
  - Removing symfony/yaml (v4.0.8)
  - Removing psr/container (1.0.0)
  - Removing pimple/pimple (v3.2.3)
  - Removing padraic/phar-updater (v1.0.6)
  - Removing padraic/humbug_get_contents (1.1.2)
  - Removing infection/infection (0.8.1)
  - Installing symfony/polyfill-ctype (v1.8.0): Loading from cache
  - Updating symfony/filesystem (v4.0.8 => v4.0.10): Downloading (100%)         
  - Updating symfony/finder (v4.0.8 => v4.0.10): Loading from cache
```

For phpstan:
```
  - Removing phpstan/phpstan (0.9.2)
  - Removing phpstan/phpdoc-parser (0.2)
  - Removing ocramius/package-versions (1.3.0)
  - Removing nette/utils (v2.5.1)
  - Removing nette/robot-loader (v3.0.3)
  - Removing nette/php-generator (v3.0.4)
  - Removing nette/neon (v2.4.2)
  - Removing nette/finder (v2.4.1)
  - Removing nette/di (v2.4.11)
  - Removing nette/bootstrap (v2.4.5)
  - Removing jean85/pretty-package-versions (1.1)
```

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -16,7 +16,7 @@ jobs:
   include:
     - stage: Static analysis (src) with phpstan
       php: 7.2
-      script: vendor/bin/phpstan analyse --level=7 src
+      script: make phpstan
```
Comment (Ocramius) [inline .travis.yml]: Please just put the executed code here, no intermediate build tools

File: .travis.yml [hunk 2]
```diff
@@ -37,7 +37,7 @@ jobs:
 
     - stage: Mutation Tests
       php: 7.2
-      script: vendor/bin/infection
+      script: make infection
```
Comment (Ocramius) [inline .travis.yml]: Same: to be reverted to bare bash

File: Makefile [hunk 3]
```diff
@@ -0,0 +1,17 @@
+PHONY: infection phpstan
+./infection.phar:
+	wget https://github.com/infection/infection/releases/download/0.8.1/infection.phar
+	chmod a+x ./infection.phar
+
+./infection.phar.pubkey:
```
Comment (Ocramius) [inline Makefile]: Until all phars are completely scoped, using phar distributions is out of question due to autoloaded dependency conflicts

Comment (BackEndTea) [reply Makefile]: Infection, in the versions below `0.9.0` does not autoload any code.

**Comment by BackEndTea** [comment]
> Hmm... not sure what this adds, what is the benefit here?

It reduces issues like the ones seen in #62 
@sebastianbergmann did a pretty good job explaining it in this thread: https://twitter.com/s_bergmann/status/999635212723212288

File: .travis.yml (hunk 1 above)
Comment (BackEndTea) [reply .travis.yml]: Would it be okay to still use `make ./phpstan.phar` to download the phar? or do you want the full wget here as well? That could cause differences in the versions used for local development and the version used on travis.

**Comment by Ocramius** [comment]
That thread is flawed because it doesn't consider static type collisions,
which happen with all the tools currently in use .

File: .travis.yml (hunk 1 above)
Comment (Ocramius) [reply .travis.yml]: I'd probably rather have a `composer.lock` handle that, heh

**Comment by asgrim** [comment]
> > Hmm... not sure what this adds, what is the benefit here?
> 
> It reduces issues like the ones seen in #62 
> @sebastianbergmann did a pretty good job explaining it in this thread: https://twitter.com/s_bergmann/status/999635212723212288

This is very interesting, and yes it's a good solution, but I have a big "BUT":

> A downside of this is that composer update won't have an effect on these dependencies. Meaning one would have to manually update their version in the makefile.

That's problematic for local development if we have to _separately_ download deps like infection/phpstan... (yes, I do run these tools locally). The collisions seen in #62 are indeed unfortunate, but they don't always happen... and it's only temporary. This seems to suggest instead that we manually install tools locally.

TL;DR - I can see benefits to both sides (thank you for bringing this to our attention!) but my personal vote would be to keep these tools managed by `composer`...

**Comment by BackEndTea** [comment]
> That's problematic for local development if we have to separately download deps like infection/phpstan... 

Currently your standard work flow would be:
```bash
$ composer install
$ vendor/bin/infection
```
Using the makefile it would now be:
```bash
$ composer install
$ make infection
```
I added the `Makefile` to ease the installation of the tools, so you don't have to manually install these tools.

The updating workflow however now does change. Simply running `composer update` will not update these dependencies, that has to be done by hand.

You could also consider using a tool like https://github.com/phar-io/phive for that.

**Comment by Ocramius** [comment]
I'm still firmly set on using composer dependencies. PHAR distributions will only be an acceptable way of sharing dependencies once https://github.com/humbug/php-scoper is working for all projects (that we rely upon) that are being distributed as PHARs.

**Comment by BackEndTea** [comment]
For what its worth: the upcoming version of infection will be fully scoped

**Comment by Ocramius** [comment]
Closing here: we'll have to wait for fully scoped PHARs before being able to move forward, and I'm generally against using PHARs for dependencies (any of them), as the SAT process is a critical/vital part of the setup and of the entire SemVer ecosystem. 

In a library that (hopefully can) proud itself to push the boundaries of BC break detection, I think it is a better idea to give out a signal to consumers that dependency constraints, resolution and version upgrades should be automated via a repository (composer's), so even if the PHARs improve, I'd rather like to have a dependency on a composer package that represents the scoped deliverable, rather than a file fetched from a random location.
